### PR TITLE
Add striped content node B-tree bucket database impl

### PR DESCRIPTION
**Pull request by vekterli**
@geirst please review. Not production ready yet due to distribution bit enforcement limitations (and more low-level testing needed), but should be good to go for benchmarking.

### CODE REVIEW

File: storage/src/vespa/storage/bucketdb/const_iterator.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include <cstdint>
+
+namespace storage::bucketdb {
+
+template <typename ConstRefT>
+class ConstIterator {
```
Comment (geirst) [inline storage/src/vespa/storage/bucketdb/const_iterator.h:9]: Consider adding class comment.

File: storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include "btree_lockable_map.h"
+#include <memory>
+#include <vector>
+
+namespace storage::bucketdb {
+
+// Bucket database implementation that stripes all superbuckets across
```
Comment (geirst) [inline storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.h:10]: Nitpick: Consider `/** */` for class comments as used elsewhere.

File: storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+namespace {
+
+constexpr uint8_t used_bits_of(uint64_t key) noexcept {
+    return static_cast<uint8_t>(key & 0b11'1111ULL);
+}
+
+}
+
+// TODO rename to sharded_btree_lockable_map instead?
```
Comment (geirst) [inline storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp]: I think we can remove this.

File: storage/src/vespa/storage/common/content_bucket_db_options.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include <cstdint>
+
+namespace storage {
+
+// Type-safe encapsulation of any options that can be passed from config
```
Comment (geirst) [inline storage/src/vespa/storage/common/content_bucket_db_options.h:8]: Nitpick: Consider using `/** */` instead.

**Review by geirst — APPROVED** [review]
👍

**Comment by geirst** [comment]
@toregge please have a look as well.

File: storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp (hunk 3 above)
Comment (vekterli) [reply storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp]: Already removed in a subsequent commit 🙂

File: storage/src/vespa/storage/bucketdb/const_iterator.h (hunk 1 above)
Comment (vekterli) [reply storage/src/vespa/storage/bucketdb/const_iterator.h:16]: Done

File: storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.h (hunk 2 above)
Comment (vekterli) [reply storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.h]: Done

File: storage/src/vespa/storage/common/content_bucket_db_options.h (hunk 4 above)
Comment (vekterli) [reply storage/src/vespa/storage/common/content_bucket_db_options.h]: Done

File: storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+template <typename T>
+StripedBTreeLockableMap<T>::~StripedBTreeLockableMap() = default;
+
+template <typename T>
+size_t StripedBTreeLockableMap<T>::stripe_of(key_type key) const noexcept {
+    assert(used_bits_of(key) >= _n_stripe_bits);
+    // Since bucket keys have count-bits at the LSB positions, we want to look at the MSBs instead.
+    return (key >> (64 - _n_stripe_bits));
```
Comment (toregge) [inline storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp:42]: This might not work as intended if _n_stripe_bits is 0.

**Review by toregge — APPROVED** [review]

File: storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp (hunk 5 above)
Comment (vekterli) [reply storage/src/vespa/storage/bucketdb/striped_btree_lockable_map.hpp:42]: `_n_stripe_bits` is asserted to be in the range [1, 8] at construction time, so this invariant should be maintained

**Pull request merged**